### PR TITLE
[HttpKernel] Add new method for getting a possible ProxyDumper

### PR DESCRIPTION
**Pull request by Spea**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | yes (well only kind of, actually it is only a refactoring)
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

In order to use my own implementation of the `ProxyDumper` in my projects, I currently have to override the `dumpContainer` method where I just copy 99% of the code and then replace only this line:
```php
$dumper->setProxyDumper(new ProxyDumper());
```

As this is not really sufficient and sometimes even breaks when the `dumpContainer` method was adjusted, I created a new method which returns the proxy dumper and can be overridden in child classes.

### CODE REVIEW

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 1]
```diff
@@ -729,8 +730,9 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
         // cache the container
         $dumper = new PhpDumper($container);
 
-        if (class_exists('ProxyManager\Configuration') && class_exists('Symfony\Bridge\ProxyManager\LazyProxy\PhpDumper\ProxyDumper')) {
-            $dumper->setProxyDumper(new ProxyDumper());
+        $proxyDumper = $this->getProxyDumper();
+        if (null !== $proxyDumper) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: can be put on one line:
`if (null !== $proxyDumper = $this->getProxyDumper()) {`

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 2]
```diff
@@ -758,6 +760,21 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
         $cache->write($rootCode, $container->getResources());
     }
 
+    /**
+     * Returns a proxy dumper implementing the
+     * {@see \Symfony\Component\DependencyInjection\LazyProxy\PhpDumper\DumperInterface} DumperInterface or null.
+     *
+     * @return DumperInterface|null
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: the full dobcblock should be removed as it doesn't provide anything more than the signature.

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 3]
```diff
@@ -758,6 +760,21 @@ protected function dumpContainer(ConfigCache $cache, ContainerBuilder $container
...
+     */
+    protected function getProxyDumper(): ?DumperInterface
+    {
+        if (class_exists('ProxyManager\Configuration') && class_exists('Symfony\Bridge\ProxyManager\LazyProxy\PhpDumper\ProxyDumper')) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: lets' use `use` statements

**Review by nicolas-grekas — COMMENTED** [review]
I'm good with the proposal once, see comments.

**Comment by Spea** [comment]
@nicolas-grekas Addressed your comments

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
@Spea I don't like the idea of exposing the proxy dumper in the Kernel. The kernel is not really concerned about this. Moreover, in 99% of the cases, you should probably not override the proxy dumper. I understand that you have a specific use case, and I think that copy/paste of the method is the most reasonable thing to do.

**Comment by Spea** [comment]
@fabpot While I get that you probably should not override the ProxyDumper in 99% of the cases I'm not sure I get your point of not extracting it to its own method. It does not affect existing code and if you are one of the 1% users who have an own implementation of the `ProxyDumper` you can override it if you want.

After all there is an `DumperInterface`, which seems a bit "useless" (besides programming design of course) if you can not use a different implementation than the one which exists in symfony itself.

However, it is probably an edge case I have here and if you think it is not worth merging, I can live with overriding the `dumpContainer` method instead.

**Comment by fabpot** [comment]
Adding a method is like adding an extension point (which should be maintained, supported, and documented). Right now, the proxy dumper is hidden, and I think we need to keep it that way. Let's close.

**Pull request closed without merging**